### PR TITLE
Show apps in header

### PR DESCRIPTION
**Pull request by juliusknorr**
This is a experimental implementation of what we've discussed with @jancborchardt @eppfel and @skjnldsv to make apps accessible directly without a popover. See https://github.com/juliushaertl/direct_menu/issues/51

I'm still not totally happy with this, but I guess we should have some discussion so please have a look,  @nextcloud/designers

### Screenshots (updated)

Up to 8 apps:
2017-03-03-151204_458x97_scrot

More than 8 apps:
2017-03-03-131705_491x96_scrot

2017-03-03-131718_586x153_scrot

2017-03-03-131727_446x230_scrot

Mobile:
2017-03-03-151025_342x168_scrot

### ToDo

- [x] Show 8 apps in total
- [x] Mobile/small screens
- [x] Cleanup CSS code to make use of scss
- [x] Fix css formating issues
- [x] Fix positioning in mobile view
- [x] Hide popover app names when popover menu is shown
- [x] Fix popover positioning after window resize
- [x] Fix app name position on small window sizes
- [x] Adding apps doesn't work any more

### CODE REVIEW

**Comment by raimund-schluessler** [comment]
What happens when I open Notes in your example? Will there be an indicator that I am using Notes anywhere?

**Comment by skjnldsv** [comment]
I like it a lot!

**Comment by MorrisJobke** [comment]
> Up to 3 apps

But this is "Up to 4 apps" - > otherwise the first 3 apps are shown plus the menu opener ;)

**Comment by juliushaertl** [comment]
> But this is "Up to 4 apps" - > otherwise the first 3 apps are shown plus the menu opener ;)
￼￼ 2￼ 
:see_no_evil: Changed it. I was not counting the apps management as an app.

**Comment by juliushaertl** [comment]
> What happens when I open Notes in your example? Will there be an indicator that I am using Notes anywhere?

@raimund-schluessler I would prefer to show the icon somwhere, not only an indicator at the 3 dots menu. Does anyone have an idea how we could do that?

**Comment by MariusBluem** [comment]
>  would prefer to show the icon somwhere, not only an indicator at the 3 dots menu. Does anyone have an idea how we could do that?

Maybe by showing the current app as the first 🤔 @juliushaertl

**Comment by Espina2** [comment]
I don't think this are a good solution. It breaks after the third/four app. I talked with @jancborchardt at the Berlin conference that maybe the best solution is to have navigation on the right. 

https://codepen.io/nikhil/pen/sicrK (example)
example (example not public) from another project Im working on

Why I think its best?

Its versatile and works on mobile. 
You can list more apps.
You can even change the order and personalize. 

I am with a little time to make a proper mockup so I hope that I can explain what I mean.

**Comment by skjnldsv** [comment]
I would be fine with that too :)

**Comment by juliushaertl** [comment]
> I talked with @jancborchardt at the Berlin conference that maybe the best solution is to have navigation on the right.

@Espina2 We also had some discussion about that over at https://help.nextcloud.com/t/is-this-the-most-awesome-or-what/6163/3

I'll just quote @jancborchardt on this:
> Yes … I was more thinking about integrating it to the left navigation without losing horizontal space (compare iCloud.com) but there are just too many drawbacks. Mainly the issue of the responsiveness and how to handle it on mobile where the navigation just hides.

**Comment by skjnldsv** [comment]
Oh common guys, i barely have the time to follow everything on github. What can I do if you discuss on nextcloud forums now? 😆

**Comment by Espina2** [comment]
This is not new even before you launch the version with icons on the topbar I have called attention to @jancborchardt at conference that in my opinion this is not a good solution. 

And like you quote 

> Yes … I was more thinking about integrating it to the left navigation without losing horizontal space (compare iCloud.com) but there are just too many drawbacks. Mainly the issue of the responsiveness and how to handle it on mobile where the navigation just hides.

The solution that I am suggesting don't add extra space. You only remove the icon for the second menu entry and add them to the first level. 

So you have icon + label on the first "menu" and only label on the second, because it's not needed the icon.

Tooltips is a crime in User Experience, if you need them its because the information or the interface isn't clear enough. For not talking about mobile that don't have "hover" capabilities. 

The solution that you are working now, its patching a problem not resolving them. This is my opinion.

**Comment by eppfel** [comment]
> Tooltips is a crime in User Experience

🚨  Call the UX Police 😂 

I think @Espina2 proposal could work. I agree that app names are not necessary as long as we have distinct icons. Problem could be external sites. We should add an option to upload a custom icon there.
And I don't see problems with the mobile version. This could be integrated into the side-menu as well.

**Comment by Spartachetto** [comment]
I am not at all an UX expert, and I get that a single user feedback and preferences have a really limited usefulness.
Yet I really like @Espina2 proposal! :-)

**Comment by SonRiab** [comment]
I would stick with the current default behavior when accessing the site with a mobile phone. Pressing on the current app name is easier then hitting small icons. But using icons and hiding the current app name for the desktop view is totally fine for me. Maybe we could add the icon in front of the app name in mobile view.
nav1
nav2

**Comment by juliushaertl** [comment]
> Tooltips is a crime in User Experience, if you need them its because the information or the interface isn't clear enough. 

I think they work quite well, especially for a menu, that you use more often. At the beginning you might need some information about what the icons will do, but after some time you'll start to remember the

> For not talking about mobile that don't have "hover" capabilities.

I would like to keep the current menu for mobile, as requires no screen space while not using it. Just as @SonRiab proposed.

> It breaks after the third/four app.

I don't like that very much either, it was just an idea to avoid a too crowded header bar.

> The solution that I am suggesting don't add extra space. You only remove the icon for the second menu entry and add them to the first level.

Ok, that could work. Feel free to do a mockup. :+1:

**Review by jancborchardt — COMMENTED** [review]
Some code nitpicking inline, and bigger design considerations here:
- Only showing 4 apps until having the »more apps« menu is too few. Let’s increase it to 8 apps. This will also fix the issue @Espina2 mentioned. Of course on mobile or when the screen gets narrower, the menu should show accordingly to make sure there’s no overflow with the username or avatar.
- How about we layout the »more apps« menu like user menu, in a list view. Fits better rather than showing bigger icons in the dropdown than there are in the bar.
- If app from more-menu is open, don’t put it first in list. Either:
	- put it as last in the list (before the »more apps« menu icon), switching out the one which was there before temporarily
	- or just put a triangle below the 3-dot-menu to indicate it, and don’t switch places
- Fix placement issues of notifications and search on the right
- Consider opening more apps and user menu on hover, like menus on Amazon
- Again, we won’t introduce a left bar for the apps _again_ because we had that in the past and decided against it because the apps/content should get as much space as possible. We have a header already.

File: core/css/header.scss [hunk 1]
```diff
@@ -421,3 +422,106 @@
...
+#appmenu li:hover span {
+	display: inline-block;
+	white-space: nowrap;
+}
+
+#appmenu li:hover span {
+	background-color: white;
+	-webkit-border-radius: 3px;
+	-moz-border-radius: 3px;
+	border-radius: 3px;
```
Comment (skjnldsv) [inline core/css/header.scss]: Okay, I'm gonna create an issue with this because everyone keeps asking! X)

**Comment by Espina2** [comment]
@jancborchardt 

> Again, we won’t introduce a left bar for the apps again because we had that in the past and decided against it because the apps/content should get as much space as possible. We have a header already.

This is not a valid point, Like I already said this left menu will have the same space. 

I only see advantages with that approach. 

First you can navigate directly to a page in a app without open the "home" app page.

The navigation are on the same place and are more natural. Like we have now you change from context from two different locations, one on the "header" and another on the left bar. 

Can you say why this is a bad pattern or give some reason to why you think this will not work?

And I want the opinions of all the others designers here
@nextcloud/designers

**Comment by jancborchardt** [comment]
> This is not a valid point, Like I already said this left menu will have the same space.

It won’t. We already have the header bar there with room for icons. The left menu will be an additional element.

> First you can navigate directly to a page in a app without open the "home" app page.

How is this different from the approach outlined here? Maybe it’s not entirely clear from the screenshot you posted.

> The navigation are on the same place and are more natural. Like we have now you change from context from two different locations, one on the "header" and another on the left bar.
> Can you say why this is a bad pattern or give some reason to why you think this will not work?

Yes, precisely for the reason that they are in the same place. Because the icons will be confused as icons for the other navigation entries, or conflict with navigation entries where there are icons. It will be a mass of icons on the left, some with text, some without. They do not have the same relevance so they don’t need to be in the same place.

Anyhow.

A bigger issue I have here is that we discussed a lot and still didn’t change anything from the status quo. The direct menu as done by @juliushaertl is the top rated app in the store and really useful to let people know we have a bunch of apps – which is not apparent at the moment.
So we should get this change in for Nextcloud 12. @nextcloud/designers @Espina2 don’t forget one of our tenets is to be pragmatic. It’s always iterations. But right now we’re brought to a stop.

**Comment by eppfel** [comment]
I support @jancborchardt, we have to go forward here. 

The biggest issue with @Espina2 approach is the horizontal space. And the app navigation + the sidebar is not a typical multi level menu, as the sidebar can function quite differently in different apps.

The biggest issue with the current approach is the mystery meat. But at least it would not be as mysterious as it is now. We could add some height to the header and add labels beneath the app icons. But I think this should not block us implementing it.

**Comment by Espina2** [comment]
@jancborchardt @eppfel 

Like I said more than one time, the horizontal space is the same, it will not add extra space.
screen-shot-2017-02-15-at-22 02 46
What I propose is like this. Have the icon for the app on the left and the pages of the app on the right. 

This have exact the same horizontal space as before. So this issue doesn't exist at all. Can you point wich app use the bar different? Because I can't see any advantage with my approach.  

At the moment I don't have the time to make an high fidelity mockup so its the best I can do.

**Comment by eppfel** [comment]
@Espina2 
bildschirmfoto 2017-02-15 um 23 48 48
bildschirmfoto 2017-02-15 um 23 49 17
bildschirmfoto 2017-02-15 um 23 51 42

**Comment by skjnldsv** [comment]
screenshot_20170216-063607

I think he means this.

**Comment by juliushaertl** [comment]
@eppfel Do you want to take over, since you have assigned yourself? Otherwise I would continue with this over the weekend.

**Comment by jancborchardt** [comment]
@juliushaertl it would be awesome if we can work together on finishing this! :) We should get it in for Nextcloud 12. Can you resolve the conflicting files, then I’d join to fix some details too. :)

And please, let’s not get sidetracked by so much discussion. This leads to standstill.

**Comment by juliushaertl** [comment]
@jancborchardt I've rebased and fixed some more issues. I'll try to get this a bit more polished tomorrow. :see_no_evil: The code is still a bit messy. If you want to jump in, feel free to do so. I've updated the todo list in the first post as well.

**Comment by juliushaertl** [comment]
Ok, this should be ready to review now. I've also updated the screenshots in the first post.

cc @jancborchardt @skjnldsv @eppfel @nextcloud/designers

**Review by MorrisJobke — APPROVED** [review]
Tested and works in IE11, Firefox, Chrome, Safari 👍

**Comment by MorrisJobke** [comment]
@nextcloud/designers Please review

**Review by MariusBluem — APPROVED** [review]
Had no problems so far with Safari and Chrome on my instance 💪

**Review by LukasReschke — APPROVED** [review]
Works nicely

**Pull request merged**